### PR TITLE
feat: add `std/ini`

### PR DESCRIPTION
**Pull request by kt3k**
Continued from https://github.com/denoland/deno_std/pull/3214

This PR tries to add `std/int` module.

### CODE REVIEW

**Comment by iuioiua** [comment]
Is this ready for review?

**Comment by kt3k** [comment]
I think this is ready for review now.

BTW I'm unsure about `IniMap` name. Ideas/suggestions for better names are also welcome.

**Review by iuioiua — APPROVED** [review]
Great! 100% test coverage and well-formatted tests.

IMO, this implementation actually does a little too much. Having only `parse()` and `stringify()` in the public API might be sufficient for almost all use cases. The same seems to be true of other similar APIs in the Standard Library.

I suggest we do the following:
1. Make `IniMap` private and only `parse()` and `stringify()` public
2. Trim any methods that aren't used in `parse()` and `stringify()`

If, in the future, there's sufficient demand for these methods, we can add them in by looking at the commits in this PR. It'd be a shame to have all this functionality in the public API that must be maintained without first confirming that there's any demand for it.

File: ini/parse.ts
[the remark this answers is not in the document]
Comment (kt3k) [reply ini/parse.ts:1]: ParseOption is used by IniMap

File: ini/stringify.ts
[the remark this answers is not in the document]
Comment (kt3k) [reply ini/stringify.ts:1]: `StringifyOptions` is also used by IniMap

**Comment by kt3k** [comment]
> IMO, this implementation actually does a little too much. Having only parse() and stringify() in the public API might be sufficient for almost all use cases.

I disagree with this. `parse` drops comment information in .ini file, and stringify doesn't restore the comments from the original .ini file. If we only provide `parse` and `stringify`, all comments in the original .ini file are always lost when it's modified by this module. I think that is below the basic expectation to this module.

`std/jsonc` has this problem, and the feature for preserving comments is actively requested https://github.com/denoland/deno_std/issues/3502

File: ini/ini_map.ts [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+/** Options for providing formatting marks. */
+export interface FormattingOptions {
+  /** The character used to assign a value to a key; defaults to '='. */
+  assignment?: string;
+  /** Character(s) used to break lines in the config file; defaults to '\n'. Ignored on parse. */
+  lineBreak?: string;
+  /** Mark to use for setting comments; expects '#', ';', '//', defaults to '#' unless another mark is found. */
+  comment?: string;
+  /** Use a plain assignment char or pad with spaces; defaults to false. Ignored on parse. */
+  pretty?: boolean;
```
[the remark this answers is not in the document]
Comment (kt3k) [reply ini/ini_map.ts:12]: `pretty` looks worth keeping to support both styles: `a = 1` and `a=1`

[the remark this answers is not in the document]
Comment (iuioiua) [reply ini/ini_map.ts:12]: I think the added flexibility provides little to no value, but my opinion isn't strong.

File: ini/ini_map.ts [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+  /** Filter duplicate keys from INI string output; defaults to false to preserve data parity. */
+  deduplicate?: boolean;
+}
+
+/** Options for parsing INI strings. */
+export interface ParseOptions {
+  /** The character used to assign a value to a key; defaults to '='. */
+  assignment?: FormattingOptions["assignment"];
```
Comment (kt3k) [inline ini/ini_map.ts:20]: This option might be unnecessary. @aaronhuggins Have you ever seen ini file using non-`=` assignment mark?

File: ini/ini_map.ts [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+
+    yield line;
+  }
+
+  #cleanFormatting(options?: FormattingOptions): FormattingOptions {
+    return Object.fromEntries(
+      Object.entries(options ?? {}).filter(([key]) =>
+        FormattingKeys.includes(key as keyof FormattingOptions)
+      ),
+    );
```
[the remark this answers is not in the document]
Comment (kt3k) [reply ini/ini_map.ts:433]: `filterKeys` should be usable here, but it seems causing too much trouble in typing.

File: ini/ini_map.ts (hunk 2 above)
Comment (iuioiua) [reply ini/ini_map.ts:20]: I've skimmed a few documents detailing the INI format, and none mention an assignment mark other than `=`. I think it's safe for us to remove this option.

**Comment by iuioiua** [comment]
I've `IniMap()` has the ability to preserve comments, can we not extend that functionality to `parse()` and `stringify()` through some `keepComments: boolean` option?

**Comment by kt3k** [comment]
`IniMap` object and `parse()` result have completely different data structures (`IniMap` stores data as list of `type Line = LineComment | LineSection | LineValue` and `parse()` result is `Record<string, unknown>`). It's not obvious how to store comment information in `parse()` result.

**Review by iuioiua — APPROVED** [review]
Ok. That's fair enough. LGTM! Good stuff! 👍🏾

**Pull request merged**